Paper: Real-Time Digital Signal Processing Using pyaudio_helper and the ipywidgets

mwickert opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

Commit:
initial commit

Commit:
Pre-final post

Commit:
Pre_final post2

Commit:
fixed code line breaks

mpacer commented:
@mwickert Your paper does not seem to build, though I see that the error that we are running into points us in two directions. Does your paper build locally on the latest 2018 branch?

mwickert commented:
M,

Both the 247 and the 250 build just find on my MacBook system. I also push some edits to the .rst files this morning.

Mark Wickert

On May 22, 2018, at 10:51 AM, M Pacer <notifications@github.com<mailto:notifications@github.com>> wrote:


@mwickert<https://github.com/mwickert> Your paper does not seem to build, though I see that the error that we are running into points us in two directions. Does your paper build locally on the latest 2018 branch?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/398#issuecomment-391062712>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AL_KBpA1zp-838NjUG4-VSQTKFNVkOzoks5t1EIlgaJpZM4UH7NL>.

mpacer commented:
@mwickert It looks like there's an error with finding `basic_DSP_IO.pdf` could you clone your repo into a new location on your computer and tell me if it works standalone there?

Commit:
Fix filename case to match figure basic_dsp_IO.pdf

mwickert commented:
M,

I see the problem is a file name case issue. The macOS does not pick this up. I just pushed a change to the .rst file to properly refer to the figure you mention.

It was basic_DSP_IO.pdf before.

.. figure:: basic_dsp_IO.pdf
:scale: 58%
:align: center
:figclass: htb

Analog signal processing implemented using real-time DSP. :label:`basicDSPio`

Mark Wickert

On May 23, 2018, at 12:32 PM, M Pacer <notifications@github.com<mailto:notifications@github.com>> wrote:


@mwickert<https://github.com/mwickert> It looks like there's an error with finding basic_DSP_IO.pdf could you clone your repo into a new location on your computer and tell me if it works standalone there?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/398#issuecomment-391452195>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AL_KBgyUhKmYSoe9TVFEqiQbD74fHYtBks5t1atEgaJpZM4UH7NL>.

mpacer commented on papers/mark_wickert_250/mwickert_250.rst:
```diff
@@ -0,0 +1,768 @@
+
+.. _SPCommTutorial: https://github.com/mwickert/SP-Comm-Tutorial-using-scikit-dsp-comm/wiki
+
+.. figure:: pyaudio_DSP_IO.pdf
```
This is still not matching case.

Commit:
Fix filename case to match figure pyaudio_dsp_IO.pdf

mwickert commented:
M,

Sorry I did not see your response earlier.

Whoops! I made the same case error twice on two figure files that have very similar file names.  The first fix I made 10 hours ago was for the file basic_DSP_IO.pdf. ==> basic_dsp_IO.pdf. I just now realized and fixed a similar case error: pyaudio_DSP_IO.pdf ==> pyaudio_dsp_IO.pdf.

.. figure:: pyaudio_dsp_IO.pdf
:scale: 58%
:align: center
:figclass: htb

Two channel analog signal processing implemented using frame-based real-time DSP. :label:`pyaudioDSPio`

I have now checked the file naming on the remaining figure pdfs and the names in the .rst file. As I see it they now all match. The PDF file building from my side is working.

I wish I would have caught this earlier.

Mark Wickert

On May 23, 2018, at 8:41 PM, M Pacer <notifications@github.com<mailto:notifications@github.com>> wrote:


@mpacer commented on this pull request.

________________________________

In papers/mark_wickert_250/mwickert_250.rst<https://github.com/scipy-conference/scipy_proceedings/pull/398#discussion_r190451757>:

> +is done using 3.5mm stereo patch cables and small pin header adapters [3p5mm]_ to interface to the AD2.
+
+Frame-based Real-Time DSP Using the :code:`io_stream` class
+-----------------------------------------------------------
+
+The block diagram of Figure :ref:`pyaudioDSPio` illustrates the essence of this paper.
+Implementing the stucture of this figure relies upon the class :code:`DSP_io_stream` which is housed
+in :code:`sk_dsp_comm.pyaudio_helper.py`. To make use of this requires the scipy stack
+(numpy, scipy, and matplotlib), as well as [DSPComm]_ and [pyaudio]_. PyAudio is supported
+supported on all majors OSs, e.g., Windows, macOS, and Linux. The configuration varies,
+but the set-up is documented at [pyaudio]_ and SPCommTutorial_. The classes and functions
+of :code:`pyaudio_helper` are detailed in Figure :ref:`pyaudioHelperclasses`.
+
+.. _SPCommTutorial: https://github.com/mwickert/SP-Comm-Tutorial-using-scikit-dsp-comm/wiki
+
+.. figure:: pyaudio_DSP_IO.pdf


This is still not matching case.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/398#pullrequestreview-122829494>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AL_KBh3mtfHOmp3Y23V2pp6mFEqJkRltks5t1h3XgaJpZM4UH7NL>.

bmcfee commented:
:hand: volunteering to help with review on this one.  I'm compiling a list of comments on the current pdf, what's the best way to deliver them?  In-thread, or as a github review?

mwickert commented:
Brian,

I am fine working directly in GitHub as I am writing now. I am not sure what the difference between in-thread and github review?

bmcfee commented:
> I am not sure what the difference between in-thread and github review?

- In-thread: dumping my comments onto this issue.

- GH review: attached directly to lines of code/text, and linked to merge status.

This is more a question for @mpacer / the scipy committee in general, but I'm fine with either way.  The benefits of GH review are that it's easier to track individual comments and have threads for each, but I'd need permissions on the repo to do it (i think?).  The benefit of PR comments is that I can do it right now.

mwickert commented:
Since I have two papers being reviewed, and the other reviewer is using in-thread, and I have used that in the past, I would prefer it this way. My plan is to make edits to the .rst file and push the changes that you will then see in the pdf. I will also plan to shown changes in this comment window as needed.

bmcfee commented:
### General summary

This package looks great!  The examples given in the paper convincingly show how to stitch together some non-trivial, real-time DSP demonstrations with a small amount of code.  Having tried to do this kind of thing before, with mixed results, I'm looking forward to trying this package out.

The contents of the paper hit most of the main points I'd like to see in describing the functionality of the tools (exceptions listed below).  The main points where I think it could be improved are: 1) a bit more context up-front describing the specific contents of this paper, and 2) some explanation of terminology/notation for the broader scipy audience.  I think these things should be fairly easy to accomplish in the given space.  Most important, I think, would be to provide some background explaining why real-time DSP can be difficult to get right (in python), and briefly summarize how the tools described here (pyaudio_helper, DSP_IO, etc) make things easier.

### Detailed comments

I'll note these as checkboxes so you can clear them as you edit.

#### Abstract
- [x] "The focus of this poster" => "paper"?
- [x] The last sentence, "Audio applications and in a reduce bandwidth fashion,..." reads a bit awkwardly.  I suggest to rephrase it.

#### Introduction
- [x] I wouldn't expect scipy readers to immediately grasp the notational difference between continuous- and discrete-time representations `x(t)` and `x[n]`.  Maybe a footnote or otherwise explicit comment would be appropriate here.
- [ ] Similarly, it may not be obvious that A<->D are hardware operations, and thus outside the scope of Python.
- [x] "operation on the signals samples" -> "signal's", or maybe just "operation on the samples `x[n]`"
- [x] "y(t) is a continuous function time t` -> "a continuous function of time t"
- [x] "`b_k, m= 0, 1, ..., M`" -> `b_m`
- [x] "So the process follows along the lines of..." reads like a sentence fragment (dangling conjunction).  I suggest to rephrase.
- [x] "The Jupyter notebook allows all of this to happen in one place..." this seems like the key point of the paper, and IMO, would be most useful to have closer to the beginning.  (I love the sampling theorem too, but it's not critical to the contributions of this paper. :))

#### Frame-based real-time DSP
- [x] As above, I think Figure 2 should be emphasized early in the paper (perhaps merged with Figure 1) so that a reader can quickly grasp the contents of this paper.
- [x] Side comment/question: is matplotlib necessary for this to work, or are `numpy/scipy` and sk_dsp_comm sufficient?
- [x] Figure 3 occupies a substantial amount of space (rightfully so), but it is barely referenced in the text.  Since it seems like a critical component of the work here, it might be helpful to spend some time in the text going over the functionality, and how the classes are intended to work together.
- [x]  "To list the internal/external devices... found in 3" -> "found in Figure 3"
- [x] Figure 4: `iMac` -> `iMic`
- [x] "frames or chucks" -> "frames or chunks"?
- [x] "The pack and unpack ... of 2" -> "of Figure 2"
- [x] "conveniently supports frame-based digital filtering" -> referring back to equation 1 would be helpful here, for readers unfamiliar with the terminology.

##### Anatomy of a pyaudio callback function
- [x] "we first need to instantiate a DSP_io_stream object": it would be good to explain why this is necessary, ie, what this obejct does in simple terms.
- [x] "The resulting plot is shown in Figure 6": since these are absolute timing numbers, it would be good to include some details about the system on which they were computed.  Would you get substantially different results using a built-in audio device instead?
- [x] "We sees that PyAudio" -> "We see"
- [x] Figure 6: C1 and C2 are not described here.  In general, for all of these plots, it would be useful to include descriptions in the caption for what each curve represents.
- [x] "The results are compared in Figure ??" broken ref
- [x] `Pxx, F = ss.my_psd(...)` -- what is `ss`?
- [x] "The results show considerable roll-off in just the ADC path" - do you have a sense of where this comes from?  Is the DAC undoing the peaking above 17K, or is there something else going on with the capture buffer?

#### Examples
- [x] It would be helpful to provide a preamble to this section briefly summarizing each of the examples.
- [x] "lelft" -> "left"
- [x] LR gain callback does not declare `DSP_IO` as global.  I think this should apply universally?  Likewise for global'ing the widgets.
- [ ] General comment: code snippets would be easier to read if contained within listing blocks and captioned
- [x] Figure 8: caption should link left/right to C1/C2, and explicitly note that the difference in amplitude is the salient factor.
- [x] "AAssuming" -> "Assuming"
- [x] Panning example: do you need to worry about clipping / overflow when converting from float to int16?
- callback2 (Design comment, independent of the paper per se), the tic-toc pattern seems to call for a context manager pattern here to cut down on boilerplate code within the callbacks.  Maybe something to consider for future versions.  (A decorator might also be nice, if it would allow streamlining the computation in the callback and facilitate jit compilation.  Just food for thought.)
- [x]  `intercative_stream` -> `interactive_stream` in a few places
- [x] "Excellent agreement is achieved..." this sentence also seems fragmented, and could perhaps use a bit of smoothing over.
- [x] "but a listening to music" -> "but listening to music"

Commit:
Fixed label typo

Commit:
A fure in the sample .rst removed

Commit:
Major editing following reviewer improvement suggestions.

Commit:
Fixed a label typo.

mwickert commented:
Brian,

Thanks for your hard work on this review. You have made many good and valid comments.

I just pushed a large number of edits to the draft .rst file. I will now comment on what I did and what is still pending further discussion. Copying your comments to the below paragraphs, I will now make responses to them and then you can see in the new version how things play out.

### Response to General Comments

*I reworked the abstract and introduction quite a bit, hoping to capture what you were thinking of here.*

### Abstract

  "The focus of this poster" => "paper"? *Fixed*
  The last sentence, "Audio applications and in a reduce bandwidth fashion,..." reads a bit awkwardly. I suggest to rephrase it. *Fixed*

### Introduction

  I wouldn't expect scipy readers to immediately grasp the notational difference between continuous- and discrete-time representations x(t) and x[n]. Maybe a footnote or otherwise explicit comment would be appropriate here. * added more discussion*
  Similarly, it may not be obvious that A<->D are hardware operations, and thus outside the scope of Python.  *added more discussion*
  "operation on the signals samples" -> "signal's", or maybe just "operation on the samples x[n]" *fixed*
  "y(t) is a continuous function time t` -> "a continuous function of time t"
  "b_k, m= 0, 1, ..., M" -> b_m *fixed*
  "So the process follows along the lines of..." reads like a sentence fragment (dangling conjunction). I suggest to rephrase. *done*
  "The Jupyter notebook allows all of this to happen in one place..." this seems like the key point of the paper, and IMO, would be most useful to have closer to the beginning. (I love the sampling theorem too, but it's not critical to the contributions of this paper. :)) *OK, I added more to the abstract, but also want to know if you would like me to remove some content?*

### Frame-based real-time DSP

  As above, I think Figure 2 should be emphasized early in the paper (perhaps merged with Figure 1) so that a reader can quickly grasp the contents of this paper. *My desire is to start with a traditional textbook DSP I/O block diagram so that the reader can better appreciate where we have end up in the PyAudio framework and RT-DSP needs.*
  Side comment/question: is matplotlib necessary for this to work, or are numpy/scipy and sk_dsp_comm sufficient? *whoops on me, yes matplotlib is needed, in fact one of the `DSP_io_stream` methods creates a plot.*
  Figure 3 occupies a substantial amount of space (rightfully so), but it is barely referenced in the text. Since it seems like a critical component of the work here, it might be helpful to spend some time in the text going over the functionality, and how the classes are intended to work together. * I introduce the figure the same, but then spread out discussion as the text rolls on, but making reference to this Figure repeatedly.*
  "To list the internal/external devices... found in 3" -> "found in Figure 3" *fixed*
  Figure 4: iMac -> iMic *fixed*
  "frames or chucks" -> "frames or chunks"?  *now discussed in the text with more detail.*
  "The pack and unpack ... of 2" -> "of Figure 2" *fixed and now discussed in the text with more detail.*
  "conveniently supports frame-based digital filtering" -> referring back to equation 1 would be helpful here, for readers unfamiliar with the terminology. *done*

### Anatomy of a pyaudio callback function

  "we first need to instantiate a DSP_io_stream object": it would be good to explain why this is necessary, ie, what this obejct does in simple terms. *more detail is now provided*
  "The resulting plot is shown in Figure 6": since these are absolute timing numbers, it would be good to include some details about the system on which they were computed. Would you get substantially different results using a built-in audio device instead?  *This is open as I want to confirm with some more experiments. I think the change is little. There are low latency USB drivers (ASIO) for certain USB audio interfaces used in sound-stage (mics and guitar, etc). I while back I tried this with a more expensive interface and found that PyAudio did not like this driver.*
  "We sees that PyAudio" -> "We see" *Fixed*
  Figure 6: C1 and C2 are not described here. In general, for all of these plots, it would be useful to include descriptions in the caption for what each curve represents. *Fixed in text and the caption*
  "The results are compared in Figure ??" broken ref *fixed*
  Pxx, F = ss.my_psd(...) -- what is ss? *explained by adding import of `sk_dsp_comm.sigsys as ss` in the code block.*
  "The results show considerable roll-off in just the ADC path" - do you have a sense of where this comes from? Is the DAC undoing the peaking above 17K, or is there something else going on with the capture buffer? *This is open as I would like to do some more experiments. I in Fall 2017 I had a PLL project where the carrier signal was centered on 16 KHz, and I don't recall seeing this peak with Sabrent device.*

### Examples

  It would be helpful to provide a preamble to this section briefly summarizing each of the examples.
  "lelft" -> "left" *not found, but maybe I fixed this earlier.*
  LR gain callback does not declare DSP_IO as global. I think this should apply universally? Likewise for global'ing the widgets. *Both are global to the notebook workspace, so there is no issue that I have run into. Maybe I am not following you on this?*
  General comment: code snippets would be easier to read if contained within listing blocks and captioned *I agree and like the way the LaTeX listing package works. In my writing I am trying to avoid the use of additional LaTeX packages. If you or anyone else has an idea I would like to hear it. For now I have tried to add more text lead in before jumping into an .rst code block*
  Figure 8: caption should link left/right to C1/C2, and explicitly note that the difference in amplitude is the salient factor. *This detail has been added*
  "AAssuming" -> "Assuming" *fixed*
  Panning example: do you need to worry about clipping / overflow when converting from float to int16? *yes you are correct. I have added a dynamic range comment in the anatomy a callback section*
callback2 (Design comment, independent of the paper per se), the tic-toc pattern seems to call for a context manager pattern here to cut down on boilerplate code within the callbacks. Maybe something to consider for future versions. (A decorator might also be nice, if it would allow streamlining the computation in the callback and facilitate jit compilation. Just food for thought.) *I agree and yes more work is needed, especially in what I call the instrumentation part of the class. I will be staying for the code sprints and would love some help on anything to the sk_dsp_comm project*
  intercative_stream -> interactive_stream in a few places *fixed*
  "Excellent agreement is achieved..." this sentence also seems fragmented, and could perhaps use a bit of smoothing over. *Fixed and more detail provided*
  "but a listening to music" -> "but listening to music" *fixed*

### Summary

Again thanks for your comments. I hope I have moved the paper in the right direction.

Commit:
Fixed figure legend swap

Commit:
Edits to comment on latency compare and gain flatness compare.

himito commented:
# General Summary

This is an interesting work since it shows how to use the interactivity of Jupyter notebooks with real-time audio processing :). Moreover, the author wrote an excellent introduction to DSP :).  **However, the paper has a lot of typos and spelling mistakes that often the desire to continue reading diminishes or disappears :(. Moreover, it is clear that the author sent the paper with no reading proof and at the last minute.**

In the following I will detail some errors. @bmcfee mentioned before some of them and the author has already fixed it :).

# Detailed comments

## Abstract
- [ ] poster -> paper
- [ ] change software defined by software-defined. 

## Introduction

- [ ] Introduce `pyaudio`, `DSP_io_stream `, `scikit-dsp-comm`. You mentioned but they are not introduced before.

## Analog Input/Output Using DSP Algorithms

- [ ] repeated "tell us"
- [ ] I don't understand the sentence: Another practical matter is in reality x[n] is a scaled and finite precision version of x(t). Is it missing a punctuation ?
- [ ] I don't understand the sentence: The DAC **does not simply set** y(nT), as y(t) is a continuous function time t.
- [ ] bk,m, m= 0,1,..,M ? do you mean bm ? or k,k=0,1,...,K ?
- [ ] you mix cellphone, cell phone. Choose one :)
- [ ] I think is better to write the URL in the same sentence of the package, e.g., DSPComm (https://github....). The same with pyaudio, ScipySignal etc. Otherwise, the reading is broken :(

## Frame-Based Real-Time DSP Using io_stream class

- [ ] What io_stram class is ? You used in a subtitle but is not introduced before.
- [ ] supported is twice
- [ ] you can replace "all majors OSs e.g., ..... by "it is multi-platform"
- [ ] Reference to "2" (three times or more), add Figure 2 please.
- [ ] "micro controller" by "microcontroller"
- [ ] "call back" by "callback"

### Performance Measurements

- [ ] I don't understand "We can also examine the callback latency by having the AD2 input a low duty cycle pulse train have a 2 Hz rate". Is it missing a punctuation?
- [ ] "We sees" by "we see"
- [ ] "and" (twice)
- [ ] There is a Figure ?? 
- [ ] The code before the figure 7 is not described. You can omit it since it only plots the file "Loop_trough_noise_SA.csv" 

## Examples

- [ ] You can recall to the reader what you will show in the examples. 

### Left and Right Gain Sliders

- [ ] "untilized" does not exist
- [ ] "gain slider" by "gain sliders"
- [ ] "coedntrol" does not exist
- [ ] "lelft" does not exist
- [ ] replace all "screen shot" by "screenshot"
- [ ] Figure 8. Add in the caption what is the blue and yellow signal.

### Cross Left-Right Channel Panning

- [ ] "AAsumming"
- [ ] Introduce "L_out", "L_in", "R_out", "R_in"
- [ ] I don't understand: "This app is best experienced by listening, but in picture form Figure 9 shows a series of scope captues". Is there a Type and it is missing a punctuation ?
- [ ] Figure 9. It's better if you split the figure 9b into 3 figures.
- [ ] "a automation" by "an automation"

### FIR and IIR Filters

- [ ] You only use a FIR filter, where do you use the IIR filter ?
- [ ] You don't describe the parameter of the filter. The arguments of the `fir_d.fir_remez_bpf`
- [ ] The caption of Fig. 10 is wrong.

### Three Band Equalizer

- [ ] The caption of Fig. 11 is wrong.
- [ ] "on" twice
- [ ] "Hz" twice

# Questions

- We can define callbacks with no `global variables`
- I think is better if you add captions to snippets of code.

bmcfee commented:
I just took a second pass over the paper, and it's looking great!  A few responses inline:


> My desire is to start with a traditional textbook DSP I/O block diagram so that the reader can better appreciate where we have end up in the PyAudio framework and RT-DSP needs.

That makes sense.



> OK, I added more to the abstract, but also want to know if you would like me to remove some content?

I think the changes to the abstract summarize it nicely.



> Both are global to the notebook workspace, so there is no issue that I have run into. Maybe I am not following you on this?

It was more a question of consistent styling in the examples, I suppose.  Some global variable (IIR state) are explicitly marked `global`, and others (`DSP_IO`) are not.  I'm fine with it either way, just checking to make sure this all works as advertised. :)



> In my writing I am trying to avoid the use of additional LaTeX packages. If you or anyone else has an idea I would like to hear it. For now I have tried to add more text lead in before jumping into an .rst code block

This seems like a question for the scipy proceedings folks.  @mpacer? 



> I agree and yes more work is needed, especially in what I call the instrumentation part of the class. I will be staying for the code sprints and would love some help on anything to the sk_dsp_comm project

Great!  (I won't be at the sprints, but I'm keen to keep up on this and help if possible.)


Two more typos in my second pass:

- `Sabarent` => `Sabrent`
- `amd` => `and`

Commit:
Reworked to make figure into parts (a)-(d).

mwickert commented:
@himito @bmcfee ,

I was just finishing up on @himito when the second round arrived from @bmcfee , so the major focus on this reply and push will be to the former, except for the typos of the latter.

**From @himito my response is integrated with the original comment text below:**

Thank you very much for your very detailed review. I have tried to address all of your comments as completely as possible. Since making changes I have started to creep over 8 pages. So, I have had two remove/compress some content, but I don't think the trimming reomved truly important content. I have not yet implemented the URL fix you suggest. I need to understand better your suggestion and will explain more when I get there in the below. 

## Abstract

  poster -> paper: *fixed*
  change software defined by software-defined. *fixed*

## Introduction

  Introduce pyaudio, DSP_io_stream, scikit-dsp-comm. You mentioned but they are not introduced before. *reworked intro following other reviews comments.*

## Analog Input/Output Using DSP Algorithms

  repeated "tell us" *fixed*
  I don't understand the sentence: Another practical matter is in reality x[n] is a scaled and finite precision version of x(t). Is it missing a punctuation ? *reworked*
  I don't understand the sentence: The DAC does not simply set y(nT), as y(t) is a continuous function time t. *reworked*
  bk,m, m= 0,1,..,M ? do you mean bm ? or k,k=0,1,...,K ? *other reviewer caught this, so yes fixed.*
  you mix cellphone, cell phone. Choose one :) *I have chosen cell phone*
  I think is better to write the URL in the same sentence of the package, e.g., DSPComm (https://github....). The same with pyaudio, ScipySignal etc. Otherwise, the reading is broken :(
Frame-Based Real-Time DSP Using io_stream class *I am not exactly sure what you point is. Right now I mostly have links to the references section and then from there URL links. Are saying I just have URL links directly in the doc and forget having them as a list of references? Maybe I am missing the point entirely, so please help me understand.*

  What io_stram class is ? You used in a subtitle but is not introduced before.
  supported is twice: *I think you mean `DSP_io_stream` class. This OK now, as it is mentioned near the end of the Introduction section.*
  you can replace "all majors OSs e.g., ..... by "it is multi-platform": *done*
  Reference to "2" (three times or more), add Figure 2 please.: *Fixed*
  "micro controller" by "microcontroller": *Fixed*
  "call back" by "callback" *Fixed*

## Performance Measurements

  I don't understand "We can also examine the callback latency by having the AD2 input a low duty cycle pulse train have a 2 Hz rate". Is it missing a punctuation? *Reworked*
  "We sees" by "we see" *Fixed*
  "and" (twice) *Fixed*
  There is a Figure ?? *Fixed*
  The code before the figure 7 is not described. You can omit it since it only plots the file "Loop_trough_noise_SA.csv" *I reworked the text here since it shows how the overlay plot is produced  from the ADC path alone and the ADC-DAC path. @bmcfee  was very interested this. 

## Examples

  You can recall to the reader what you will show in the examples. *Done according too @bmcfee request*

### Left and Right Gain Sliders

  "untilized" does not exist *Fixed earlier, as I could not find it*
  "gain slider" by "gain sliders" *Fixed*
  "coedntrol" does not exist *Fixed earlier as I could not find it*
  "lelft" does not exist *Fixed earlier*
  replace all "screen shot" by "screenshot" *Done/adopted*
  Figure 8. Add in the caption what is the blue and yellow signal. *Done*

### Cross Left-Right Channel Panning

  "AAsumming" *Fixed*
  Introduce "L_out", "L_in", "R_out", "R_in" *Detail added*
  I don't understand: "This app is best experienced by listening, but in picture form Figure 9 shows a series of scope captues". Is there a Type and it is missing a punctuation ? *Reworked*
  Figure 9. It's better if you split the figure 9b into 3 figures. *To this figure is a piecewise animation, thus keeping it together is more appropriate. I have trimmed the screenshot portion of parts (a) and have made parts (b)-(d) as follows: Part (b) shows 0%, part (c) shows 50%, and part (d) shows 100%.
  "a automation" by "an automation" *Fixed*

### FIR and IIR Filters

  You only use a FIR filter, where do you use the IIR filter ? *Renamed the subsection to FIR Bandpass Filter. I originally had another example planned, but IIR is now in the equalizer example.*
  You don't describe the parameter of the filter. The arguments of the fir_d.fir_remez_bpf *I provide some detail and then add a URL to an examples page on scikit-dsp-comm*
  The caption of Fig. 10 is wrong. *Fixed*

### Three Band Equalizer

  The caption of Fig. 11 is wrong. *Fixed*
  "on" twice *Fixed*
  "Hz" twice *Fixed*

## Questions

We can define callbacks with no global variables: *Are you telling me or asking me? If it is a question the answer is yes as in globals are not required unless the algorithm need to have a state maintained from callback-to-callback. Filters in particular need to maintain state. I admit I am coming at this from the perspective of a C programmer. If these a trick in Python I am interested in knowing it.*
I think is better if you add captions to snippets of code. *The reviewer @bmcfee asked about this too and I wrote: In my writing I am trying to avoid the use of additional LaTeX packages. If you or anyone else has an idea I would like to hear it. For now I have tried to add more text lead in before jumping into an .rst code block.*

**From @bmcfee Two more typos in my second pass:**

Sabarent => Sabrent: *fixed*
amd => and: *not found, so I must have fixed it*

Commit:
Changes in response to @himito and @bmcfee reviewer comments.

mwickert commented:
@bmcfee 

To comment on your: It was more a question of consistent styling in the examples, I suppose. Some global variable (IIR state) are explicitly marked global, and others (DSP_IO) are not. I'm fine with it either way, just checking to make sure this all works as advertised. :)

I am somewhat puzzled on this. I do not have to make the DSP_IO object global, but if I fail to make a filter state global the code will not run. When I program real-time DSP in C on a micro controller and use ISRs (interrupt services routines), I have to make the C struct that holds filter coefficients and filter states global. What is Python doing here?

bmcfee commented:
> I am somewhat puzzled on this. I do not have to make the DSP_IO object global, but if I fail to make a filter state global the code will not run.

By "not run" do you mean it gets an undefined variable exception, or behaves incorrectly (does not propagate `zi`)? [This link](https://www.geeksforgeeks.org/global-local-variables-python/) might explain what's going on in terms of scoping.

My $2e-2 is that it's not wrong to declare all the global objects as `global` (including DSP_IO, widgets, etc), and the code would be clearer for it, even if it runs either way.

himito commented:
Hi @mwickert,
Thanks for taking in account my comments :)

> I think is better to write the URL in the same sentence of the package, e.g., DSPComm (https://github....). The same with pyaudio, ScipySignal etc. Otherwise, the reading is broken :(
> Frame-Based Real-Time DSP Using io_stream class I am not exactly sure what you point is. Right now I mostly have links to the references section and then from there URL links. Are saying I just have URL links directly in the doc and forget having them as a list of references? Maybe I am missing the point entirely, so please help me understand.

Yes, insted of putting them in the reference section, I think it's better if you put them directly. It will easier to read :)

mwickert commented:
@bmcfee,

I read the link you provided and do see behavior similar. If for example I omit filter states I get this error:

```python
Exception in thread Thread-14:
Traceback (most recent call last):
  File "/Users/markwickert/anaconda3/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/Users/markwickert/anaconda3/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/markwickert/Documents/Projects/scikit-dsp-comm/sk_dsp_comm/pyaudio_helper.py", line 151, in stream_thread
    self.stream(Tsec=time,numChan = channel)
  File "/Users/markwickert/Documents/Projects/scikit-dsp-comm/sk_dsp_comm/pyaudio_helper.py", line 202, in stream
    time.sleep(self.sleep_time)
UnboundLocalError
```
I agree that being consistent is best and will make the changes to all of the code examples to have all related variables global in the callback, i.e., DSP_IO, various widgets, and DSP algorithm state variables, in particular filters. I am also now planing to add a reference in the paper to a corresponding Jupyter notebook example, rendered as HTML on the project's GH-pages and also in an examples folder under the project itself. What do you and others think? 

Putting this into practice, below is a 3D audio application that positions a sound source in 3D using widgets and plays the output to headphones. All needed variables are now global in the callback:
```python
def callback(in_data, frame_length, time_info, status):
    global DSP_IO,subject,zi_left,zi_right,r_xz_plane,azimuth,y_axis,Gain
    DSP_IO.DSP_callback_tic()
    # convert byte data to ndarray
    in_data_nda = np.frombuffer(in_data, dtype=np.int16)
    # separate left and right data
    x_left,x_right = DSP_IO.get_LR(in_data_nda.astype(float32))    
    #***********************************************
    # DSP operations here:
    # Apply Kemar HRIR left and right channel filters at
    # the sound source location in cylindrical coordinates
    # mapped to cartesian coordinates
    # The input to both filters comes by first 
    # combining x_left & x_right channels
#     x_mono = Gain.value*(x_left + x_right)/2
    x_mono = Gain.value*5000*randn(frame_length) #input white noise
    subject.cart2ipcs(r_xz_plane.value*sin(pi/180*azimuth.value), #x
                      y_axis.value, #y
                      r_xz_plane.value*cos(pi/180*azimuth.value)) #z
#     subject.cart2ipcs(0,0,-1)
    y_left, zi_left = signal.lfilter(subject.coeffL,1,
                                     subject.tL*x_mono,zi=zi_left) 
    y_right, zi_right = signal.lfilter(subject.coeffR,1,
                                       subject.tR*x_mono,zi=zi_right)
    #***********************************************
    # Pack left and right data together
    y = DSP_IO.pack_LR(y_left,y_right)
    # Typically more DSP code here     
    #***********************************************
    # Save data for later analysis
    # accumulate a new frame of samples
    DSP_IO.DSP_capture_add_samples_stereo(y_left,y_right)
    #***********************************************
    # Convert from float back to int16
    y = y.astype(int16)
    DSP_IO.DSP_callback_toc()
    # Convert ndarray back to bytes
    return y.tobytes(), pah.pyaudio.paContinue

# Create a ss_mapping2CIPIChrir object
# SUBJECT 20, 21 (KEMAR sm), & 165 (KEMAR LG) available now
subject = ss_mapping2CIPIChrir('subject_165')
# Initialize L/R filter initial conditions
zi_left = signal.lfiltic(subject.coeffL,1,[0])
zi_right = signal.lfiltic(subject.coeffR,1,[0])
# Create a IO stream object and start streaming
DSP_IO = pah.DSP_io_stream(callback,0,1,frame_length=1024, 
                           fs=44100,Tcapture=0)
DSP_IO.interactive_stream(0,2)
widgets.HBox([Gain,r_xz_plane,azimuth,y_axis])
```

bmcfee commented:
> I am also now planing to add a reference in the paper to a corresponding Jupyter notebook example, rendered as HTML on the project's GH-pages and also in an examples folder under the project itself. What do you and others think?

That's a great idea.

mwickert commented:
@himito 

In response to your clarifying the URLs linking, it turns out I started writing with URLs embedded in the doc, but then thought some readers would want to see a list at the end. I can definitely make this move, and now the question: do I still keep a list of these same links in the references section for those readers that like to see that?

himito commented:
Personally, I don't think so, however, what do you think @bmcfee ?

bmcfee commented:
> Personally, I don't think so, however, what do you think @bmcfee ?
> 

My preference would be for URLs as references or footnotes, since that's the norm for academic papers in most disciplines.  Really, it should be up to the scipy proceedings [style guidelines](https://github.com/scipy-conference/scipy_proceedings#general-information-and-guidelines-for-authors), though they seem to be vague on this point.

himito commented:
The problem with footnotes is that one page could have a lot of URLs :\.
If we use URLs as references, so it is important to add more details about the URL, for example 
- Scipy documentation <https://docs.scipy.org/doc/>
and not only a list of URLs that we don't know what they are :)

bmcfee commented:
> The problem with footnotes is that one page could have a lot of URLs :.
> If we use URLs as references, so it is important to add more details about the URL,

I don't see why either of those is a problem?  APA style for websites at bare minimum would include the url, title, and date accessed.  That seems reasonable to me.

himito commented:
> APA style for websites at bare minimum would include the url, title, and date accessed. That seems reasonable to me.


Yes, also for me :)

mwickert commented:
@bmcfee @himito 

It seems you two have reached a consensus on URL, title, and date accessed. So for me what exactly would like me to me doing with my References and the URL links in the paper?

1. Keep the links in the paper as is, what I mean here is link to references first and then let the references take you outside to the URL. To follow APA I should have these references have additionally the title and the date accessed. Shall I go for all three?

2. Do something else like direct URL links in the paper, which I am not as fond of?

3. Wait for input from the committee?

Commit:
Adjustment to figure labels

Commit:
Include consistent use of globals.

dalippa commented:
@mwickert I think we should do citations with APA (http://pitt.libguides.com/citationhelp)

@mpacer @FatihAkici @deniederhut : Fatih and I decided we'll use this format at today's meeting

Commit:
Change to APA style references made.

mwickert commented:
@dalippa,

The changes have been pushed. Shall I do likewise on my other paper?

dalippa commented:
@mwickert yes, please do use APA on the other paper too.

deniederhut commented:
Hi @himito and @bmcfee ! Thanks for your work reviewing this paper. Do you feel it is now ready to be included in the proceedings?

bmcfee commented:
LGTM!

himito commented:
@deniederhut yeah ! :D

Commit:
Fixed a broken reference link.

Commit:
Fix typo in conclusions.
